typed-th-in-ghc-9/article.md

```diff
@@ -0,0 +1,205 @@
+
+The `template-haskell` package was changed in version 2.17.0.0 and GHC version 9.0 according to the [Make Q (TExp a) into a newtype](https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0195-code-texp.rst) proposal, in which the typed expression quasi-quoter (`[|| ... ||]`) now returns a different datatype.
+
+Under the new specification, typed quotations now return `Quote m => Code m a` instead of `Q (TExp a)`. This is a breaking change for existing code, and codebases targeting GHC 9.0 either need to adapt their TTH code or use an alternative approach for compatibility.
+
+The `Code` newtype is simply a wrapper around an `m (TExp a)`, defined similarly to this:
+
+```hs
+newtype Code m a = Code (m (TExp a))
+```
+
+The actual definition is a bit more complicated, as it contains some levity-polymorphism mechanisms. If you go to [its definition](https://hackage.haskell.org/package/template-haskell-2.18.0.0/docs/src/Language.Haskell.TH.Syntax.html#Code), you will actually see this:
+
+```hs
+type role Code representational nominal
+newtype Code m (a :: TYPE (r :: RuntimeRep)) = Code
+  { examineCode :: m (TExp a) -- ^ Underlying monadic value
+  }
+```
+
+The differences between the simplified and real definitions are not important for the understanding of this post, and you may keep the simplified definition in mind while reading through this.
+
+Thankfully, the untyped TH interface remains mostly unchanged, and the typed TH interface should only need to use `Code` instead of `TExp` to compile again with some helper functions.
+
+<!-- TODO: Write about motivation as well? -->
+
+## Example GHC 8 code
+
+To make the examples here easier to follow, let us define a small typed Template Haskell module that we wish to work under GHC 8, which we will later port to GHC 9. The code will be quite simple, its purpose is to parse an environment flag into a known data type or stop the compilation altogether if it fails.
+
+We will use the [`template-haskell`](https://hackage.haskell.org/package/template-haskell) package, so don't forget to load it. You will also need to activate the `TemplateHaskell` language extension.
+
+```hs
+{-# LANGUAGE TemplateHaskell #-}
+
+module TH where
+
+import Control.Monad.IO.Class (liftIO)
+import Language.Haskell.TH.Syntax (Q, TExp)
+import System.Environment (lookupEnv)
+
+data LogLevel
+  = Development  -- ^ Log errors and debug messages
+  | Production  -- ^ Only log errors
+  deriving (Show)
+
+logLevelFromFlag :: Q (TExp LogLevel)
+logLevelFromFlag = do
+  flag <- liftIO $ lookupEnv "LOG_LEVEL"
+  case flag of
+    Nothing -> [|| Production ||]  -- We default to Production if the flag is unset
+    Just level -> case level of
+      "DEVELOPMENT" -> [|| Development ||]
+      "PRODUCTION" -> [|| Production ||]
+      other -> fail $ "Unrecognized LOG_LEVEL flag: " <> other
+```
+
+If you try to compile this module with GHC 9, however, this example will fail to compile:
+
+```hs
```

Comment:
For a minute, I was pretty confused thinking that you meant the `text` package and I couldn't understand where it should be used. Now I see you meant using `text` syntax instead of `hs`. I've changed it.